[ADD] Developer: data module tutorial (#11050)
odoo/documentation: content/developer/tutorials/importable_modules.rst

sandboxed environment.

This environment contains several utilities to help you interact with the Odoo database:

- ``self``: the record on which the action is executed
- ``env``: the environment of the record
- ``model``: the model of the record
- ``user`` and ``uid``: the current user and their id
- ``datetime``, ``dateutil``, ``timezone`` and ``time``: libraries to help with date/time calculations
- ``float_compare``: a utility function to compare two float values with a given precision
- ``b64encode`` and ``b64decode``: utility functions to encode and decode values in base64
- ``Command``: a utility class to help build complex expressions and commands (see the `Command`
  class in the :ref:`ORM reference <reference/fields/relational>`)

In addition, you have access to the recordset on which the action is executed (typically a single
record when the action is executed from a form view, and multiple records when the action is
executed from a list view) via the ``record`` and ``records`` variables.

.. note::

    If your action needs to return an action to the client (for example to redirect the user to

>     another view), you can assign it to a semi-magical ``action`` variable inside your server

Reviewer: Not super reassuring :D
```suggestion
    another view), you can assign it to an ``action`` variable inside your server
```

    html += '</ul></body></html>'
    response = request.make_response(html)
        ]]></field>
    </record>

Several useful methods are available in the `request` object to facilitate the generation of the
response object:

- `request.render(template, qcontext=None, lazy=True, **kw)` to render a QWeb template using its
  xmlid; the extra keyword arguments are forwarded to the `werkzeug.Response` object (e.g. to set
  cookies, headers, etc.)
- `request.redirect(location, code=303, local=True)` to redirect to a different URL; the `local`
  argument is used to specify whether the redirection should be relative to the website or not
  (default: `True`).
- `request.notfound()` to return a `werkzeug.HTTPException` exception to signal a 404 error to
  the website.
- `request.make_response(data, headers=None, cookies=None, status=200)` to manually create a
  `werkzeug.Response` object; the `status` argument is the HTTP status code to return (default:
  200).
- `request.make_json_response(data, headers=None, cookies=None, status=200)` to manually create a
  JSON response; the data will be json-serialiazed using `json.dumps` utility; this can be useful

>   to set up machine-to-machine communications via API calls.

Reviewer: machine-to-machine sounds like the Quebec version of S2S :smile:

    with an impossible domain (e.g. `[('id', '=', False)]`)
  - add a record rule so that properties marked as `x_api_published` can be read by the
    public user
  - add a server action to return a list of properties in JSON format when the URL
    `/actions/api/estate` is called

A sprinkle of Javascript
========================

Whilst importable modules cannot include Python files, no such restriction exists for Javascript
files. Adding Javascript files to your importable module is exactly the same as adding them
to a standard Odoo module.

This means that an importable module can include new field components or even entirely new views.

As an example, let's add a simple 'tour' to the Estate module. Tours are a standard mechanism in
Odoo used to onboard users by guiding them through your application.

A very minimal tour with a single step can be added by adding this file in `static/src/js/tour.js`:

.. code-block:: js

    import { registry } from "@web/core/registry";

    registry.category("web_tour.tours").add('estate_tour', {
        url: "/web",
        sequence: 170,
        steps: () => [{

>         trigger: '.o_app[data-menu-xmlid="documents.menu_root"]',

Author: wrong selector
